### PR TITLE
PowCache Support Part 2

### PR DESCRIPTION
**Pull request by CodeIsTheKey**
Restructured PowCache to remove locks, logging, and file I/O.

Moved logging and file I/O into Init.

cs_main is locked by all callers (except two, which were added).

This removes link dependencies from libconsensus.

### CODE REVIEW

File: src/index/base.cpp [hunk 1]
```diff
@@ -177,17 +177,19 @@ void BaseIndex::ThreadSync()
...
-                FatalError("%s: Failed to read block %s from disk",
-                           __func__, pindex->GetBlockHash().ToString());
-                return;
-            }
-            if (!WriteBlock(block, pindex)) {
-                FatalError("%s: Failed to write block %s to index database",
-                           __func__, pindex->GetBlockHash().ToString());
-                return;
+            {
+                LOCK(cs_main);
```
Comment (CodeIsTheKey) [inline src/index/base.cpp:181]: This is one of two places where ReadBlockFromDisk/WriteBlock were not protected by cs_main.  PowCache is expecting to be protected externally, so we need this lock.

File: src/validation.cpp [hunk 2]
```diff
@@ -4140,6 +4140,7 @@ void CChainState::LoadExternalBlockFile(FILE* fileIn, FlatFilePos* dbp)
                 std::deque<uint256> queue;
                 queue.push_back(hash);
                 while (!queue.empty()) {
+                    LOCK(cs_main);
```
Comment (CodeIsTheKey) [inline src/validation.cpp:4143]: This is one of two places where ReadBlockFromDisk was not protected by cs_main.  PowCache is expecting to be protected externally, so we need this lock.

**Comment by Bigpiggy01** [comment]
#LFG

**Review by Bigpiggy01 — APPROVED** [review]
MEWN ITTTTT!!!!!

**Review by nandofw — APPROVED** [review]

**Pull request merged**